proposals/p2240.md

```diff
@@ -0,0 +1,758 @@
+programmers who have less familiarity with variadics.
+
+This decision was made in
+[leads issue #1162](https://github.com/carbon-language/carbon-lang/issues/1162).
+
+### Use postfix instead of prefix operators
+
+`...` is a postfix operator in C++, and `[:]` is a postfix operator in C++
+proposal P1858R2. Furthermore, in both cases postfix position is arguably more
+natural: postfix `...` aligns with the natural-language usage of "…", and
+postfix `[:]` aligns with C++'s (and Carbon's) subscripting syntax. So in some
+ways it would be very desirable to make their Carbon counterparts postfix as
+well. Notably, it would let us describe variadic evaluation very cleanly in
+terms of replacing `[:]` with `[$I]`. However, this turns out to be an awkward
+fit for Carbon. Consider this example:
+
+```
+fn SumInts(..., [:]params: [i64;]) -> i64
+```
+
+If `[:]` and `...,` were suffix operators, this would instead be:
+
+```
+fn SumInts(params: [i64;][:] ...,) -> i64
+```
+
+But this is ambiguous about whether the `[:]` operator applies to
+`params: [i64;]` or just `[i64;]`, so we need to add a set of parentheses:
+
+```
+fn SumInts((params: [i64;])[:] ...,) -> i64
+```
+
+To state the point more generally: when an expression mixes prefix and suffix
+operators, their order is determined by precedence and parentheses, which is
+usually less readable than relying on the linear order of operators with the
+same fixity. The binding operator `:` is effectively a prefix operator (because
+its left-hand side is a single token), so if `[:]` is a suffix operator, code
+that mixes the two will be somewhat harder to read. That's a particular problem
+because, as our motivating examples illustrate, variadic function signatures
+will often have those two operators in close conjunction, and readability is at
+a premium in function signatures.
+
+C solved a very similar problem with mixing prefix `*` and postfix `.` by
+introducing a separate `->` operator. We could in principle solve this issue the
+same way, introducing a separate "variadic binding operator" `$`, such that
+`params$ [i64;]` is equivalent to `(params: [i64;])[:]`. However, this would
+trade a readability problem for a teachability problem, by introducing more
+syntax that programmers need to learn before they can even read variadic
+function signatures.
+
+### Keyword syntax
+
+> **TODO:** Decide whether to make this the primary proposal, and restructure
+> the discussion as appropriate.
+
+We could build the syntax around keywords rather than symbolic tokens, much like
+the
+[Swift proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0393-parameter-packs.md).
+For concreteness, the keywords I will discuss here are as follows:
+
+| Symbolic token | Keyword replacement               |
+| -------------- | --------------------------------- |
+| `[:]`          | `expand`                          |
+| `...,`         | `repeat` and `unpack` (see below) |
```

Comment:
I believe what was proposed in the discussion in person was `pack` instead of `unpack`

Reply:
> take the scrutinee tuple, pack it, and then match it against the `params` binding pattern

That was my reading, and seems natural in that a pattern is generally a description of how to match the scrutinee.

The different semantics of `pack` come from being used in a pattern context, where the arity is determined by the scrutinee rather than what follows the word `pack`.

Reply:
From discussion with @chandlerc: one issue that arises with pack parameters is whether the name is singular or plural. Singular works better for uses of the parameter, but (absent some suitable introductory keyword or similar) plural seems to work better when introducing the parameter. And even with the keyword `pack`, plural seems to work better than singular.

One option that we discussed and seems to potentially be promising is using `each` both when introducing the pack and when referencing the elements of it:

```
fn F[..., each T:! type](..., each x: T) {
  ...{
    Print(each x);
  }
}
```

This both fixes the plurality problem and doesn't have the verb / noun problems that `pack` has.

Reply:
Yeah, it should, fixed.